File: python code/quickSort.py
import random
import time
import psutil


def quick_sort(arr, low, high):
    if low < high:
        partition = low - 1
        pivot = arr[high]
        for i in range(low, high):
            if arr[i] <= pivot:
                partition += 1
                arr[partition], arr[i] = arr[i], arr[partition]
        arr[partition + 1], arr[high] = arr[high], arr[partition + 1]

        quick_sort(arr, low, partition)
        quick_sort(arr, partition + 2, high)

# main
if __name__ == "__main__":

  # Get random sample size for testing
    size = int(input("Enter a size: "))
    array = [random.randint(0, 99) for _ in range(size)]

    start_time = time.time()

    # the function call
    quick_sort(array, 0, size - 1)

    stop_time = time.time()

    # changing the seconds to micoseconds
    duration = (stop_time - start_time) * 1e6  

    # Print the execution time, CPU time, and memory usage
    print("Merge sort time is:", duration, "microseconds")

    # Get cpu usuage
    cpu_usage = psutil.cpu_percent()
    print("CPU usage is:", cpu_usage, "%")

    # Get memory usage
    memory_usage = psutil.virtual_memory().used / 1024 
    print("Memory usage:", memory_usage, "KB")

    print("Sorted array:")
    print(array)

File: C++ Code/For Linux os/quickSort.cpp
#include <iostream>
#include <vector>
#include <chrono>
#include <fstream>
#include <sstream>
#include <unistd.h> 

using namespace std::chrono;
using namespace std;

// Function to measure CPU usage
double getCPUUsage() 
{
    ifstream file("/proc/stat");
    string line;
    getline(file, line);
    istringstream iss(line);
    string cpu;
    int user, nice, system, idle, iowait, irq, softirq, steal, guest, guest_nice;
    iss >> cpu >> user >> nice >> system >> idle >> iowait >> irq >> softirq >> steal >> guest >> guest_nice;
    int idle_time = idle + iowait;
    int non_idle_time = user + nice + system + irq + softirq + steal;
    int total_time = idle_time + non_idle_time;
    return ((non_idle_time / static_cast<double>(total_time)) * 100); 
}

long getMemoryUsage() 
{
    ifstream file("/proc/self/status");
    string line;
    long memoryUsageKB = 0;
    while (getline(file, line)) 
    {
        istringstream iss(line);
        string key;
        if (iss >> key) 
        {
            if (key == "VmRSS:") 
            { 
                long value;
                if (iss >> value) 
                {
                    memoryUsageKB = value;
                    break; 
                }
            }
        }
    }
    return memoryUsageKB;
}

void swap(int &x, int &y)
{
    int temp;
    temp = x;
    x = y;
    y = temp;
}

void quickSort(vector<int> &arr, int low, int high)
{
    if (low < high)
    {
        int partition = 0;
        int pivot = arr[high];
        int temp = (low - 1);
        for (int i = low; i <= high - 1; i++)
        {
            if (arr[i] <= pivot)
            {
                temp++;
                swap(arr[temp], arr[i]);
            }
        }
        swap(arr[temp + 1], arr[high]);
        partition = (temp + 1);

        quickSort(arr, low, partition - 1);
        quickSort(arr, partition + 1, high);
    }
}

int main() {
    // Get random sample size
    vector<int> array;
    int size;
    cout << "Enter a size: ";
    cin >> size;
    for (int i = 0; i < size; i++) {
        array.push_back(rand() % 100);
    }

    // Getting the time of start for the execution
    auto start = high_resolution_clock::now();

    quickSort(array, 0, size - 1);


    // Getting the time of completion for the execution
    auto stop = high_resolution_clock::now();
    auto duration = duration_cast<microseconds>(stop - start);

    // Print the results
    cout << "Quick sort time: " << duration.count() << " microseconds" << endl;
    cout << "CPU usage: " << getCPUUsage() << "%" << endl;
    cout << "Memory usage: " << getMemoryUsage() << " KB" << endl;

    // Print the sorted list
    for (int j = 0; j < size; j++) {
        cout << array[j] << ' ';
    }
    cout << endl;

    return 0;
}
